# pandas
pandas is a fast, powerful, flexible and easy to use open source **data analysis and manipulation tool**, built on top of the Python programming language.

pandas makes working with “relational” or “labeled” data both easy and intuitive.

pandas blends the high-performance, array-computing ideas of NumPy with the flexible data manipulation capabilities of spreadsheets and relational databases (such as
SQL). 

### What Is Data Science?
There’s a joke that says a data scientist is someone who knows more
statistics than a computer scientist and more computer science than a
statistician

### NumPy
NumPy, short for Numerical Python, has long been a cornerstone of numerical computing in Python. 
It provides the data structures, algorithms, and library glue needed for most scientific applications involving numerical data in Python. 
NumPy contains, among other things:
- A fast and efficient multidimensional array object `ndarray` (The "nd" in ndarray stands for "n-dimensional")
- Functions for performing element-wise computations with arrays or mathematical operations between arrays
- Tools for reading and writing array-based datasets to disk
- Linear algebra operations, Fourier transform, and random number generation
- A mature C API to enable Python extensions and native C or C++ code to access NumPy’s data structures and computational facilities

One of the reasons NumPy is so important for numerical computations in Python is because it is designed for **efficiency on large arrays of data**. There are a number of reasons for this:
- NumPy internally stores data in a contiguous block of memory, independent of other built-in Python objects. NumPy’s library of algorithms written in the C lan‐
guage can operate on this memory without any type checking or other overhead. NumPy arrays also use much less memory than built-in Python sequences.
- NumPy operations perform complex computations on entire arrays without the need for Python for loops.

While pandas adopts many coding idioms from NumPy, the biggest difference is that pandas is designed for working with tabular or **heterogeneous data**. 


## Install
pandas can be installed via pip from PyPI.

`pip install pandas`

### Import Conventions
The Python community has adopted a number of naming conventions for commonly used modules:

In [113]:
import pandas as pd
import numpy as np

In [155]:
# Random dataframe generation

common_names = ['Adam', 'Alexander', 'Amelia', 'Andrew', 'Anna', 'Benjamin', 'Caroline', 'Charlotte', 'Daniel', 'David', 'Dylan', 'Elizabeth', 'Ella', 'Emily', 'Emma', 'Ethan', 'Grace', 'Hannah', 'Isabella', 'Jack', 'Jacob', 'James', 'John', 'Joseph',
                        'Joshua', 'Katherine', 'Lauren', 'Liam', 'Lily', 'Madison', 'Matthew', 'Mia', 'Michael', 'Natalie', 'Nicholas', 'Olivia', 'Rachel', 'Robert', 'Samuel', 'Sarah', 'Sophia', 'Thomas', 'Tyler', 'Victoria', 'William', 'Xavier', 'Zachary', 'Zoe', 'Zoey', 'Aiden']

random_users = pd.DataFrame(columns=["Name", "Gender", "Social Media", "Education Level", "Salary", "Is Active","Country"])

for i in range(1000):
    name = np.random.choice(common_names)
    gender = np.random.choice(["Female", "Male"])
    social = np.random.choice(["Facebook", "Instagram", "Twitter", "Reddit", "WeChat", "TikTok", "Youtube"])
    edu_level = np.random.choice(["High School", "Bachelor's Degree", "Master's Degree", "PhD"])
    country = np.random.choice(["USA", "Canada", "UK", "Australia", "Germany", "France", "Japan", "China", "India", "Russia"])
    salary = np.random.randint(0, 1000)
    active = np.random.choice(["Yes", "No"])
    
    random_users.loc[i] = [name, gender, social, edu_level, salary, active, country]

print(random_users)

          Name  Gender Social Media    Education Level  Salary Is Active   
0     Nicholas    Male       WeChat  Bachelor's Degree     961       Yes  \
1       Thomas    Male       WeChat  Bachelor's Degree     483       Yes   
2       Robert  Female     Facebook        High School       6       Yes   
3      Matthew    Male       TikTok                PhD       1       Yes   
4     Caroline    Male       TikTok        High School     757        No   
..         ...     ...          ...                ...     ...       ...   
995      Tyler    Male     Facebook  Bachelor's Degree     489       Yes   
996      James  Female    Instagram                PhD     515       Yes   
997     Andrew  Female       WeChat  Bachelor's Degree     534        No   
998        Zoe  Female      Twitter    Master's Degree     163        No   
999  Alexander    Male       Reddit        High School     193        No   

       Country  
0        Japan  
1    Australia  
2      Germany  
3       Canada  
4 

<!--<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">-->
<style>
.alert-danger {
  color: #721c24;
  background-color: #f8d7da;
  border-color: #f5c6cb;
  padding: 0.75rem 1.25rem;
  margin-bottom: 1rem;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  margin-left: auto;
  margin-right: auto;
  max-width: 800px;
}

.alert-secondary {
  color: #383d41;
  background-color: #e2e3e5;
  border-color: #d6d8db;
  padding: 0.75rem 1.25rem;
  margin-bottom: 1rem;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  margin-left: auto;
  margin-right: auto;
  max-width: 800px;
}

.mark{
        color: #e83e8c;
        font-weight: 400;
        font-size: 87.5%;
        line-height: 1.65;
        word-wrap: break-word;
        box-sizing: border-box;
        font-family: "SFMono-Regular",Menlo,Consolas,Monaco,Liberation Mono,Lucida Console,monospace;

        padding: 0.1rem 0.25rem;
        background-color: #e1e1e1;
        border: 1px solid #f5f5f5;
        border-radius: 0.25rem;
    }

</style>
# pandas Datastructures

## Important Notice

<div class="alert-danger">
  Many Pandas operations have an `inplace` parameter, defaulting to `False`, meaning the original DataFrame is untouched, and the operation returns a new DataFrame.
  <br><br>Using the <span class="mark">inplace=True</span> in a pandas method changes the default behaviour -  the operation on the dataframe returns None (doesn’t return anything), it instead ‘modifies the underlying data’. It mutates the actual object which you apply it to.


</div>

```python
        # Whether to modify the DataFrame rather than creating a new one. 
        inplace: bool, default False
```

Why is inplace = False the default behavior?
- There is a discussion that most in-place and out-of-place versions of a method create a copy of the data anyway, thus in general, there are no performance benefits.
- Since inplace=True returns None, it does not work with method chaining.
```python
    result = df.some_function1().reset_index().some_function2()
```

Notice the difference between calling a method that has an inplace parameter:

```python
    # returns a new DataFrame, therefore the original should be assigned to it.
    df = df.set_index('ISO2')

    # returns none, mutates df. Therefore no need for assignment. (df=)
    df.set_index('ISO2', inplace=True)
```
</div>

## Index
pandas’s Index objects are responsible for holding the axis labels and other metadata (like the axis name or names). 
Index objects are immutable and thus can’t be modified by the user.
DataFrame is a tabular, column-oriented <font color="Coral">data structure with both row (index) and column labels</font>

## A.Series
A Series is a one-dimensional array-like object containing a sequence of values (of similar types to NumPy types) and an associated array of data labels, called its index.
(Like a column in the table)

Note that the visual display of a Series is just **plain text**, as opposed to the nicely styled table for DataFrames. You will also see the data type or dtype of the Series.

### Creating Series

1. Series can be instantiated from a list:
```python
    calling_codes = [90, 44, 1, 81, 86]
    pd.Series(calling_codes)

```
- Since we did not specify an index for the data, a default one consisting of the integers 0 through N-1:
```
    0    90
    1    44
    2     1
    3    81
    4    86
    dtype: int64
```
- We can specify index in the constructor:
```python
    calling_codes = [90, 44, 1, 81, 86]
    labels = ['TR', 'UK', 'US', 'JP', 'CN']
    pd.Series(calling_codes, index=labels)

```
```
    TR    90
    UK    44
    US     1
    JP    81
    CN    86
    dtype: int64
```

- The passed index is a list of axis labels.

2. Series can be instantiated from dicts:
```python
    calling_codes = {'TR': 90, 'UK': 44, 'US': 1, 'JP': 81, 'CN': 86}
    pd.Series(calling_codes)
```
- Note that the dictionary already is the combination of labels and calling_codes lists above.
```
    TR    90
    UK    44
    US     1
    JP    81
    CN    86
    dtype: int64
```





In [12]:
# Series can be instantiated from a list:
# Since we did not specify an index for the data, a default one consisting of the integers 0 through N-1 

calling_codes = [90, 44, 1, 81, 86]
pd.Series(calling_codes)

0    90
1    44
2     1
3    81
4    86
dtype: int64

In [13]:
# Series can be instantiated from a list:
# We can specify an index:

calling_codes = [90, 44, 1, 81, 86]
labels = ['TR', 'UK', 'US', 'JP', 'CN']
pd.Series(calling_codes, index=labels)

TR    90
UK    44
US     1
JP    81
CN    86
dtype: int64

In [14]:
# Series can be instantiated from dicts:

calling_codes = {'TR': 90, 'UK': 44, 'US': 1, 'JP': 81, 'CN': 86}
pd.Series(calling_codes)

TR    90
UK    44
US     1
JP    81
CN    86
dtype: int64

## B. DataFrame
When working with tabular data, such as data stored in spreadsheets or databases, pandas is the right tool for you. pandas will help you to explore, clean, and process your data. 

In pandas, a data table is called a **DataFrame**.

Lets say we have country data like this:
```
  ISO2  code            name     capital
  --------------------------------------
    TR    90          Turkey      Ankara
    UK    44  United Kingdom      London
    US     1   United States  Washington
    JP    81           Japan       Tokyo
    CN    86           China     Beijing
```


### Creating DataFrame

The syntax of creating dataframe is:
```python
  pandas.DataFrame(data, index, columns)
```
- `data`: dataset from which dataframe is to be created. It can be list, dictionary, scalar value, series, ndarrays, etc.
- `index`: optional, by default the index of the dataframe starts from 0 and ends at the last data value(n-1). It defines the row label explicitly.
- `columns`: array-like parameter, used to provide Column labels in the dataframe. Defaulting to RangeIndex(0, 1, 2, …, n)

0. Using data as a list:

- The list will be a column in the frame, where the columnname is specified with the `columns` parameter:
```python
      data = [10,20,30,40,50,60]
      df = pd.DataFrame(data, columns=['Numbers'])
```
1. Using a dictionary
  ```python
        country = {
          "ISO2"    : "TR",
          "call"    : 90,
          "name"    : "Turkey",
          "capital" : "Ankara"
        }
        df = pd.DataFrame(country)
  ```

2. Using 2D List:
  - regular 2d lists can be visualized as a matrix - a list of rows, where is row is a list.
  - note that data itself has no information about axis labels.
```python
[
    [90, 'Turkey', 'Ankara'], 
    [44, 'United Kingdom', 'London'], 
    [1, 'United States', 'Washington'], 
    [81, 'Japan', 'Tokyo'], 
    [86, 'China', 'Beijing']
]
```
3. Using List of dictionaries [{}, {}, {}]
  - a single row in a table is like a JSON representation of an object - a dictionary.
  - in each dictionary, have information about column names, which are the <font color="Coral">keys</font>.
  

  
  4. Create DataFrame from a dictionary, Columns mapped to lists
   - For the plural form, think about the same JSON, but this time, the value is not a scalar - its a list.
   ```python
        countries = {
          "ISO2"    : ["TR", "UK", "US", "JP", "CN"]
          "call"    : [90, 44, 1, 81, 86]
          "name"    : ["Turkey", "United Kingdom", "United States", "Japan", "China"]
          "capital" : ["Ankara", "London", "Washington DC", "Tokyo", "Beijing"]
        }
  ```

  5. Create DataFrame from a dictionary, Columns mapped to dicts with row labels as keys.
   - Notice the difference btw 4: The column to be used as index (ISO2) is removed, and its values are used as a key in the dictionaries:
  ```python
        countries = {
          'call'    : {'TR': 90, 'UK': 44, 'US': 1, 'JP': 81, 'CN': 86},
          'name'    : {'TR': 'Turkey', 'UK': 'United Kingdom', 'US': 'United States','JP': 'Japan', 'CN': 'China'},
          'capital' : {'TR': 'Ankara', 'UK': 'London', 'US': 'Washington DC','JP': 'Tokyo', 'CN': 'Beijing'}
        }
  ```

 6. Create DataFrame from a dictionary, Rows mapped to dicts with column labels as keys.
   - We can think of a row as a heterogenous data structure, with a combination of columns identifed by an index (PK, row label)
   ```python
        {
          'TR': {'call': 90, 'name': 'Turkey','capital': 'Ankara'},
          'UK': {'call': 44, 'name': 'United Kingdom','capital': 'London'},
          'US': {'call': 1, 'name': 'United States','capital': 'Washington DC'},
          'JP': {'call': 81, 'name': 'Japan','capital': 'Tokyo'},
          'CN': {'call': 999, 'name': 'China','capital': 'Beijing'}
        }
   ```




In [15]:
# Example 0 - Create DataFrame form a list:
data = [10, 20, 30, 40]

# Option 1 - default row and column labels
df = pd.DataFrame(data)
""" 
    0
0  10
1  20
2  30
3  40
"""

# Option 2 - assign row and column labels using index and columns parameters while constructing df:
df = pd.DataFrame(data, index=['ten', 'twenty', 'thirty', 'fourty'], columns=['ID'])
""" 
        ID
ten     10
twenty  20
thirty  30
fourty  40
"""

print(df)

        ID
ten     10
twenty  20
thirty  30
fourty  40


In [41]:
# Example 1 - Create a DataFrame form a dictionary:

country = {
    "ISO2": "TR",
    "call": 90,
    "name": "Turkey",
    "capital": "Ankara"
}

df = pd.DataFrame(country, index=[0,])

# note that the column labels are dictionary keys:
""" 
   ISO2  call    name capital
0    TR    90  Turkey  Ankara
"""

print(df)

  ISO2  call    name capital
0   TR    90  Turkey  Ankara


In [16]:
# Example 2 - Create DataFrame from 2D List:

""" 
    0               1           2
0  90          Turkey      Ankara
1  44  United Kingdom      London
2   1   United States  Washington
3  81           Japan       Tokyo
4  86           China     Beijing
"""

countries = [[90, 'Turkey', 'Ankara'], [44, 'United Kingdom', 'London'], [1, 'United States', 'Washington'], [81, 'Japan', 'Tokyo'], [86, 'China', 'Beijing']]
df = pd.DataFrame(countries)

# print df
print(df)

    0               1           2
0  90          Turkey      Ankara
1  44  United Kingdom      London
2   1   United States  Washington
3  81           Japan       Tokyo
4  86           China     Beijing


In [17]:
# Note that previous the previous df had default integer labels for columns 
# If you specify a sequence of columns, the DataFrame’s columns will be arranged in that order (columns=)
# Often it will be desirable to create a Series with an index identifying each data point with a label (index=)
country_codes = pd.DataFrame(
    countries,
    index=['TR', 'UK', 'US', 'JP', 'CN'],
    columns=['call', 'name', 'capital']
)

# print df
print(country_codes)

    call            name     capital
TR    90          Turkey      Ankara
UK    44  United Kingdom      London
US     1   United States  Washington
JP    81           Japan       Tokyo
CN    86           China     Beijing


In [18]:
# Example 3.1 - Create DataFrame from a List of dictionaries:
# Notice that each item is a row in the frame.
# Note that in each dictionary, have information about column names, which are the 'keys'.
country1 = { "ISO2": "TR", "call": 90, "name": "Turkey", "capital": "Ankara"}
country2 = { "ISO2": "UK", "call": 44, "name": "United Kingdom", "capital": "London"}
country3 = { "ISO2": "US", "call": 1, "name": "United States", "capital": "Washington DC"}
country4 = { "ISO2": "JP", "call": 81, "name": "Japan", "capital": "Tokyo"}
country5 = { "ISO2": "CN", "call": 86, "name": "China", "capital": "Beijing"}

countries = [country1, country2, country3, country4, country5]
df = pd.DataFrame(countries)

# print df
print(df)


  ISO2  call            name        capital
0   TR    90          Turkey         Ankara
1   UK    44  United Kingdom         London
2   US     1   United States  Washington DC
3   JP    81           Japan          Tokyo
4   CN    86           China        Beijing


In [19]:
# Example 3.2 - Create DataFrame from a List of dictionaries with an index

# Create a list of dictionaries representing the grades of 10 students
students = [
    {'Name': 'Harry Potter', 'Charms': 85, 'Defense Against the Dark Arts': 90, 'Herbology': 75, 'Potions': 80, 'Transfiguration': 90},
    {'Name': 'Ron Weasley', 'Charms': 75, 'Defense Against the Dark Arts': 80, 'Herbology': 70, 'Potions': 85, 'Transfiguration': 75},
    {'Name': 'Hermione Granger', 'Charms': 95, 'Defense Against the Dark Arts': 100, 'Herbology': 90, 'Potions': 95, 'Transfiguration': 100},
    {'Name': 'Draco Malfoy', 'Charms': 80, 'Defense Against the Dark Arts': 85, 'Herbology': 65, 'Potions': 75, 'Transfiguration': 85},
    {'Name': 'Luna Lovegood', 'Charms': 90, 'Defense Against the Dark Arts': 95, 'Herbology': 80, 'Potions': 90, 'Transfiguration': 95},
    {'Name': 'Neville Longbottom', 'Charms': 70, 'Defense Against the Dark Arts': 75, 'Herbology': 95, 'Potions': 70, 'Transfiguration': 75},
    {'Name': 'Ginny Weasley', 'Charms': 85, 'Defense Against the Dark Arts': 90, 'Herbology': 80, 'Potions': 85, 'Transfiguration': 90},
    {'Name': 'Cedric Diggory', 'Charms': 90, 'Defense Against the Dark Arts': 95, 'Herbology': 85, 'Potions': 90, 'Transfiguration': 95},
    {'Name': 'Cho Chang', 'Charms': 95, 'Defense Against the Dark Arts': 90, 'Herbology': 80, 'Potions': 85, 'Transfiguration': 90},
    {'Name': 'Seamus Finnigan', 'Charms': 75, 'Defense Against the Dark Arts': 80, 'Herbology': 70, 'Potions': 75, 'Transfiguration': 80}
]

# Create a pandas dataframe from the list of dictionaries
df = pd.DataFrame(students)

# Set the 'Name' column as the index
df.set_index('Name', inplace=True)

# Print the dataframe
print(df)


                    Charms  Defense Against the Dark Arts  Herbology  Potions   
Name                                                                            
Harry Potter            85                             90         75       80  \
Ron Weasley             75                             80         70       85   
Hermione Granger        95                            100         90       95   
Draco Malfoy            80                             85         65       75   
Luna Lovegood           90                             95         80       90   
Neville Longbottom      70                             75         95       70   
Ginny Weasley           85                             90         80       85   
Cedric Diggory          90                             95         85       90   
Cho Chang               95                             90         80       85   
Seamus Finnigan         75                             80         70       75   

                    Transfi

In [38]:
# Example 4 - Create DataFrame from a dictionary, Columns mapped to lists
# Note that we are using pandas.DataFrame() constructor.
# For the plural form, think about the same JSON, but this time, the value is not a scalar - its a list.
# Note that this dictionary has information about columns, but no information about index (no row labels): default integer index from 0 to n-1
# Note that PK column (ISO2) is just a regular column, not specified to be an index.
countries = {
     "ISO2"    : ["TR", "UK", "US", "JP", "CN"],
     "call"   : [90, 44, 1, 81, 86],
     "name"    : ["Turkey", "United Kingdom", "United States", "Japan", "China"],
     "capital" : ["Ankara", "London", "Washington DC", "Tokyo", "Beijing"]
}

# Note that we are using pandas.DataFrame() constructor.
df = pd.DataFrame(countries)  # default index, 0 to n-1

# Set the 'ISO2' column as the index
df = df.set_index('ISO2')
# df.set_index('ISO2', inplace=True)


# print df
df

call            name        capital
ISO2                                     
TR      90          Turkey         Ankara
UK      44  United Kingdom         London
US       1   United States  Washington DC
JP      81           Japan          Tokyo
CN      86           China        Beijing

In [ ]:
# Example 5 - Create DataFrame from a dictionary, Columns mapped to dicts with row labels as keys
# Note that we are using DataFrame.from_dict() constructor.  (orient="columns")
# We can think of a row as a heterogenous data structure, with a combination of columns identifed by an index (PK, row label)
# Note that we have both columns and index - columns are keys, and index values are the inner keys. (row labels).

countries = {
    'call': {'TR': 90, 'UK': 44, 'US': 1, 'JP': 81, 'CN': 86},
    'name': {'TR': 'Turkey', 'UK': 'United Kingdom', 'US': 'United States','JP': 'Japan', 'CN': 'China'},
    'capital': {'TR': 'Ankara', 'UK': 'London', 'US': 'Washington DC','JP': 'Tokyo', 'CN': 'Beijing'},
}

# Note that we are using DataFrame.from_dict() constructor
df = pd.DataFrame.from_dict(countries, orient="columns")

# print df
print(df)

    call            name        capital
TR    90          Turkey         Ankara
UK    44  United Kingdom         London
US     1   United States  Washington DC
JP    81           Japan          Tokyo
CN    86           China        Beijing


In [ ]:
# Example 6 - Create DataFrame from a dictionary, Rows mapped to dicts with column labels as keys.
# Note that we are using DataFrame.from_dict() constructor.  (orient="index")
# We can think of a row as a heterogenous data structure, with a combination of columns identifed by an index (PK, row label)
# Note that we have both columns and index - index are keys, and column values are the inner keys.

countries = {
    'TR': {'call': 90, 'name': 'Turkey','capital': 'Ankara'},
    'UK': {'call': 44, 'name': 'United Kingdom','capital': 'London'},
    'US': {'call': 1, 'name': 'United States','capital': 'Washington DC'},
    'JP': {'call': 81, 'name': 'Japan','capital': 'Tokyo'},
    'CN': {'call': 999, 'name': 'China','capital': 'Beijing'}
}

# note .from_dict() constructor
df = pd.DataFrame.from_dict(countries, orient="index")  # look at the dictionary above, each row is a row of frame as well, the orientation is index (row)

# print df
print(df)

    call            name        capital
TR    90          Turkey         Ankara
UK    44  United Kingdom         London
US     1   United States  Washington DC
JP    81           Japan          Tokyo
CN   999           China        Beijing


In [ ]:
# Example: Create DataFrame from a dictionary (plural form)

lastnames = ['Booker', 'Grey', 'Johnson', 'Jenkins', 'Smith']
emails = ['bo@example.com', 'gr@example.com', 'jo@example.com', 'je@example.com', 'sm@example.com']
usernames = ['booker12', 'grey07', 'johnson81', 'jenkins46', 'smith79']

# A dictionary where keys are the "column names" and values are the lists:
users_dict = {'LastName': lastnames, 'Email': emails, 'Username': usernames}

# Create a DataFrame from a Python dict:
df_users = pd.DataFrame(users_dict)

print(df_users)


  LastName           Email   Username
0   Booker  bo@example.com   booker12
1     Grey  gr@example.com     grey07
2  Johnson  jo@example.com  johnson81
3  Jenkins  je@example.com  jenkins46
4    Smith  sm@example.com    smith79


In [ ]:
# Example: Create DataFrame from a dictionary (plural form)

scores_gryffindor = {
    'Name': ['Ron', 'Harry', 'Hermione'],
    'Math': [65, 60, 69],
    'Sci': [65, 60, 69],
    'Hist': [65, 60, 69],
    'Econ': [65, 60, 69]
}

df_scores_gryffindor = pd.DataFrame(scores_gryffindor)
df_scores_gryffindor.set_index('Name')
print(df_scores_gryffindor)


       Name  Math  Sci  Hist  Econ
0       Ron    65   65    65    65
1     Harry    60   60    60    60
2  Hermione    69   69    69    69


In [ ]:
# Example: Create DataFrame from a dictionary, Row Index (orient="index")

scores_gryffindor_rowindexaskey = {
    'Harry': {
        'Math': 60,
        'Sci': 70,
        'Hist': 80,
        'Econ': 90
    },
    'Ron': {
        'Math': 65,
        'Sci': 75,
        'Hist': 85,
        'Econ': 95
    },
    'Hermione': {
        'Math': 69,
        'Sci': 79,
        'Hist': 89,
        'Econ': 99
    }
}

df = pd.DataFrame.from_dict(scores_gryffindor_rowindexaskey, orient='index')
print(df)


          Math  Sci  Hist  Econ
Harry       60   70    80    90
Ron         65   75    85    95
Hermione    69   79    89    99


## Importing Data
pandas features a number of functions for reading tabular data as a DataFrame object.  (csv, excel, sql, json, parquet,…). How to work with these data sources is provided by function with the prefix:
- import data: `read_*`
- store data : `to_*` 

```python
    df = pd.read_csv("data/users.csv")
```

- We can specify the column which contains row labels (index):
```python
    df = pd.read_csv("data/brics.csv", index_col=0)
```


### Working with missing data 
Handling missing values is an important and frequently nuanced part of the file parsing process. Missing data is usually either not present (empty string) or marked by some sentinel value. By default, pandas uses a set of commonly occurring sentinels,
such as NA and NULL

The `pd.isnull()` and `pd.notnull()` functions in pandas should be used to detect missing data.
Series also has these as instance methods, ie. seriesname.isnull()


In [ ]:
df = pd.read_csv("data/users.csv")
print(df)

       Username                   Email    Id First name Last name Active
0      blu3FisH      rachel@example.com  9012     Rachel    Booker    Yes
1     GreenFr0g   laura@yourcompany.com  2070      Laura      Grey    Yes
2      blackd0g   craig@yourcompany.com  4081      Craig   Johnson     No
3     jenkins46        mary@example.com  9346       Mary   Jenkins    Yes
4       smith79       jamie@example.com  5079      Jamie     Smith    Yes
5        john00               js@co.com   303       John     Smith     No
6   BlackTurkey    jhalprin@example.com   304        Jim   Halprin    Yes
7      BlueHawk      tjones@example.com   305     Teresa     Jones    Yes
8     GreenTree    tomjones@example.com   306      Tommy     Jones    Yes
9    OrangeFish  greggjones@example.com   307      Gregg     Jones    Yes
10      RedBoat   dthompson@example.com   308     Daniel  Thompson    Yes


In [ ]:
# We can specify the column which contains row labels (index):
brics = pd.read_csv("data/brics.csv", index_col=0)
print(brics)

           country    capitol    area  population
code                                             
BR          Brazil   Brasilia   8.516      200.40
RU          Russia     Moscow  17.100      143.50
IN           India  New Delhi   3.286     1252.00
CH           China    Beijing   9.597     1357.00
SA    South Africa   Pretoria   1.221       52.98


In [ ]:
# Missing values
# str: missing(NaN), space(empty), "" (NaN)
# numerics: missing(NaN), space(empty), negative
df = pd.read_csv("data/missing_values.csv")
# print(brics)

# boolean array representing isnull of a column:
# null_lastnames = pd.isnull(brics['Lastname']) # built-in function
null_lastnames = df['Lastname'].isnull()    # object method

df[null_lastnames]   # filter by column indexing

Desc Lastname Quiz  Midterm  Final Grade
0  Empty string      NaN   70       70     70     B
1       Missing      NaN   77       77     77     B
2   Sentinel NA      NaN   77       77     77     B

### head(), tail() and info()
To view a small sample of a Series or DataFrame object, use the head() and tail() methods. The default number of elements to display is five, but you may pass a custom number.

In [ ]:
df.head()

Desc Lastname Quiz  Midterm  Final Grade
0  Empty string      NaN   70       70     70     B
1       Missing      NaN   77       77     77     B
2   Sentinel NA      NaN   77       77     77     B
3    space char            88       88     88     A
4       Missing  Alfalfa  NaN       80     90     A

In [ ]:
df.tail(3)

Desc Lastname Quiz  Midterm  Final Grade
7          zero     Mike    0       50     50     D
8      negative   Alfred  -60       60     80    60
9  sentinel NaN     John  NaN       50     50   NaN

In [ ]:
brics.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, BR to SA
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   country     5 non-null      object 
 1   capitol     5 non-null      object 
 2   area        5 non-null      float64
 3   population  5 non-null      float64
dtypes: float64(2), object(2)
memory usage: 200.0+ bytes


### Attributes and underlying data



`shape`: gives the axis dimensions of the object, consistent with ndarray
(entries, columns)

In [ ]:
brics.shape

(5, 4)

In [ ]:
brics.columns

Index(['country', 'capitol', 'area', 'population'], dtype='object')

### Selecting Columns
The Python and NumPy indexing operators [] and attribute operator . provide quick and easy access to pandas data structures across a wide range of use cases.

A column in a DataFrame can be retrieved as a Series either by dict-like notation or by attribute:

- Selecting a single column as a Series 
<br>To select a single column of data, simply put the name of the column in-between the brackets.
```
brics['capital']
brics.capital:
```

In [ ]:
brics['capitol']    # Series

code
BR     Brasilia
RU       Moscow
IN    New Delhi
CH      Beijing
SA     Pretoria
Name: capitol, dtype: object

You can get the array representation and index object of the Series via its `values` and `index` attributes, respectively:

In [ ]:
type(brics['capitol'])  # pandas.core.series.Series


pandas.core.series.Series

In [ ]:
brics['capitol'].index      # Index(['BR', 'RU', 'IN', 'CH', 'SA'], dtype='object', name='code')


Index(['BR', 'RU', 'IN', 'CH', 'SA'], dtype='object', name='code')

In [ ]:
brics['country'].values     # array(['Brazil', 'Russia', 'India', 'China', 'South Africa'], dtype=object)

array(['Brazil', 'Russia', 'India', 'China', 'South Africa'], dtype=object)

Selecting multiple columns as a DataFrame: `brics[['country', 'capital']]`
<br>Pass a list as an argument to df index: [ [] ]

In [ ]:
brics[['country', 'capitol']]

country    capitol
code                         
BR          Brazil   Brasilia
RU          Russia     Moscow
IN           India  New Delhi
CH           China    Beijing
SA    South Africa   Pretoria

In [ ]:
type(brics[['country', 'capitol']])  # pandas.core.frame.DataFrame


pandas.core.frame.DataFrame

<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">

<style>
    code{
      color: red;
    }
    
    .mark{
        color: #e83e8c;
        font-weight: 400;
        font-size: 87.5%;
        line-height: 1.65;
        word-wrap: break-word;
        box-sizing: border-box;
        font-family: "SFMono-Regular",Menlo,Consolas,Monaco,Liberation Mono,Lucida Console,monospace;

        padding: 0.1rem 0.25rem;
        background-color: #e1e1e1;
        border: 1px solid #f5f5f5;
        border-radius: 0.25rem;
    }
</style>

### Selecting Row(s)
The `.loc` indexer can select subsets of rows or columns. 
Most importantly, it only selects data by the **LABEL** of the rows and columns.

```python
    df.loc[row(s), column(s)]
```

The primary function of indexing with `[]` (a.k.a. `__getitem__` for those familiar with implementing class behavior in Python) is selecting out lower-dimensional slices. The following table shows return type values when indexing pandas objects with []:

<table class="table">
  <thead>
    <tr>
      <th scope="col">Object Type</th>
      <th scope="col">Selection</th>
      <th scope="col">Return Type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <!--<th scope="row">Series</th>-->
      <td>Series</td>
      <td><span class="mark">series[label]</span></td>
      <td>scalar value</td>
    </tr>
    <tr>
      <!--<th scope="row">Series</th>-->
      <td>DataFrame</td>
      <td><span class="mark">frame[colname]</span></td>
      <td><span class="mark">Series</span> corresponding to colname</td>
    </tr>
  </tbody>
</table>

Allowed inputs are:
1. A single label              : 'TR'
2. A list or array of labels   : ['TR', 'UK']
3. A slice object with labels  : 'TR' : 'UK'
4. A boolean array             : [True, False]

#### Slicing
- Slicing with labels behaves differently than normal Python slicing in that the end‐point is inclusive.
- Note that when slicing, no braces inside .loc[:,:]

#### .loc[] Indexer Syntax:
<table>
  <thead>
    <tr>
      <th scope="col">Selection</th>
      <th scope="col">Syntax</th>
      <th scope="col">Return</th>
    </tr>
  </thead>
    <tbody>
    <tr>
      <!--<th scope="row">Series</th>-->
      <td>Select a single row</td>
      <td><span class="mark">.loc['RU']<span></td>
      <td>`Series`</td>
    </tr>
    <tr>
      <td>Select multiple rows</td>
      <td><span class="mark">.loc[['IN', 'RU']]<span></td>
      <td><code>DataFrame</code></td>
    </tr>
    <tr>
      <td>Select a range of rows</td>
      <td><span class="mark">.loc['IN':'SA']<span></td>
      <td><code>DataFrame</code></td>
    </tr>
    <tr>
      <td>Select a cell (row and column)</td>
      <td><span class="mark">.loc['RU', 'capitol']<span></td>
      <td>Scalar value</td>
    </tr>
    <tr>
      <td>Select 2 rows and 2 columns</td>
      <td><span class="mark">.loc[['IN', 'RU'],['country', 'capital']]<span></td>
      <td><code>DataFrame</code></td>
    </tr>
    <tr>
      <td>Select a column (all of the rows)</td>
      <td><span class="mark">.loc[:, ['capital']]<span></td>
      <td><code>DataFrame</code></td>
    </tr>
    <tr>
      <td>Select 2 columns (all of the rows)</td>
      <td><span class="mark">.loc[:, ['country', 'capital']]<span></td>
      <td><code>DataFrame</code></td>
    </tr>

  </tbody>
</table>

In [ ]:
brics.loc['RU']

country       Russia
capitol       Moscow
area            17.1
population     143.5
Name: RU, dtype: object

In [ ]:
brics.loc[['RU', 'IN']]


country    capitol    area  population
code                                       
RU    Russia     Moscow  17.100       143.5
IN     India  New Delhi   3.286      1252.0

In [ ]:
brics.loc['IN':'SA']    # Note that .loc includes the last value with slice notation!

country    capitol   area  population
code                                            
IN           India  New Delhi  3.286     1252.00
CH           China    Beijing  9.597     1357.00
SA    South Africa   Pretoria  1.221       52.98

In [ ]:
brics.loc['RU', 'capitol']

'Moscow'

In [ ]:
brics.loc[['IN', 'RU'],['country', 'capitol']]

country    capitol
code                   
IN     India  New Delhi
RU    Russia     Moscow

In [ ]:
# Select (all rows of) 2 columns
brics.loc[:, ['country', 'capitol']]

country    capitol
code                         
BR          Brazil   Brasilia
RU          Russia     Moscow
IN           India  New Delhi
CH           China    Beijing
SA    South Africa   Pretoria

In [ ]:
brics.loc[:, 'capitol'] # Series


code
BR     Brasilia
RU       Moscow
IN    New Delhi
CH      Beijing
SA     Pretoria
Name: capitol, dtype: object

In [ ]:
brics.loc[:, ['capitol']]
# type(brics.loc[:, ['capitol']])   # DataFrame


capitol
code           
BR     Brasilia
RU       Moscow
IN    New Delhi
CH      Beijing
SA     Pretoria

In [ ]:
brics.loc[:, 'capitol':'population']

capitol    area  population
code                               
BR     Brasilia   8.516      200.40
RU       Moscow  17.100      143.50
IN    New Delhi   3.286     1252.00
CH      Beijing   9.597     1357.00
SA     Pretoria   1.221       52.98

### Selecting subsets with .iloc
The .iloc indexer is very similar to .loc but only uses integer locations to make its selections.

pandas provides a suite of methods in order to get purely integer based indexing. (0-based)
```
df.iloc[0]
```
- Selecting a single row with .iloc: `.iloc[0]` 
- Selecting multiple rows with .iloc: `.iloc[[0, 2, 4]]`
- Use slice notation to select a range of rows: `.iloc[3:5]`
- Selecting 2 rows and 2 columns: `.iloc[[0,4], [0, 2]]`

In [ ]:
brics.iloc[0]   # Series

country         Brazil
capital       Brasilia
area             8.516
population       200.4
Name: BR, dtype: object

In [ ]:
brics.iloc[[0, 2, 4]]   # remember, don't do df.iloc[5, 2, 4]


country    capital   area  population
code                                            
BR          Brazil   Brasilia  8.516      200.40
IN           India  New Delhi  3.286     1252.00
SA    South Africa   Pretoria  1.221       52.98

In [ ]:
brics.iloc[2:4]


country    capital   area  population
code                                      
IN     India  New Delhi  3.286      1252.0
CH     China    Beijing  9.597      1357.0

In [ ]:
brics.iloc[[0, 4], [0, 2]]

country   area
code                     
BR          Brazil  8.516
SA    South Africa  1.221

## Boolean indexing
Another common operation is the use of boolean vectors to filter the data. 
- Using a boolean vector to index a Series works exactly as in a NumPy ndarray.
- The operators are: | for or, & for and, and ~ for not.

```python
    # column:
    calling_codes = {'TR':90, 'UK':44, 'US':1, 'JP':81, 'CN':86}

    # Create Series from dictionary:
    s = pd.Series(calling_codes)

```
- Note that when we write a condition on a scalar, it returns boolean value.
We can write the condition on a Series also, then it returns a boolean vector!
```python
    filter1 = s > 80    # Series[bool]
```
- On a DataFrame object, to write a condition on a series, first we get the column by index operator []:
```python
    filter2 = df['column'] > 1    # Series[bool]
```

Using a boolean vector as a filter works as a mask:
```python
    s[s>80]
```

<pre>
    TR    90            TR     True                     TR    90
    UK    44            UK    False                     JP    81
    US     1     x      US    False                 =   CN    86
    JP    81            JP     True                     dtype: int64
    CN    86            CN     True
    dtype: int64        dtype: bool
</pre>

1. Using a boolean vector to index a Dataframe: 
```python
    df[df['column'] > 1000]`
```

<pre>
                        BR    False                         country    capitol   area  population
                        RU    False                     code                                      
       df        x      IN     True                 =   IN     India  New Delhi  3.286      1252.0
                        CN     True                     CH     China    Beijing  9.597      1357.0
                        SA    False
    dtype: int64        Name: population, dtype: bool
</pre>

2. Using a boolean vector with the `.loc[]` indexer to filter a Dataframe: 
```python
    df.loc[filter, column(s)]
```

- Example: Select capitols of countries where population is over a billion:
```python
    brics.loc[brics['population'] > 1000, ['capitol']]
```

3. Using `.isin()` method to filter, just like a chain of == operators:
- Example: Select population of India and China
```python
    brics.loc[brics['country'].isin(['India', 'China']), ['population']]
```



In [ ]:
# boolean vector to index a Series:

# row:
country = {'ISO':'TR', 'call':90, 'capitol':'Ankara', 'name':'Turkey'}
# column:
call_dict = {'TR':90, 'UK':44, 'US':1, 'JP':81, 'CN':86}

# Create Series from dictionary:
calling_codes = pd.Series(calling_codes)

# print(calling_codes)

# Note that when we write a condition on a scalar, it returns boolean value.
# We can write the condition on a Series also, then it returns a boolean vector!
filter1 = calling_codes > 80    # Series[bool]

print(filter1)

""" 
TR     True
UK    False
US    False
JP     True
CN     True
dtype: bool
"""

# Using a boolean vector to index a Series:
df_filtered = calling_codes[calling_codes > 80]

print(df_filtered)
""" 
TR    90
JP    81
CN    86
dtype: int64
"""

x = 0   # to stop last """ """ from printing

TR     True
UK    False
US    False
JP     True
CN     True
dtype: bool
TR    90
JP    81
CN    86
dtype: int64


In [ ]:
filter = brics['population'] >= 1000.0

high_pop = brics[filter]
high_pop_country = brics.loc[filter, 'country']

print(high_pop)
print(high_pop_country)


     country    capitol   area  population
code                                      
IN     India  New Delhi  3.286      1252.0
CH     China    Beijing  9.597      1357.0
code
IN    India
CH    China
Name: country, dtype: object


In [ ]:
# Select population of India and China
brics.loc[brics['country'].isin(['India', 'China']), ['population']]

population
code            
IN        1252.0
CH        1357.0

In [ ]:
# Select capitols of countries where population is over a billion:
brics.loc[brics['population'] > 1000.0, ['capitol']]

capitol
code           
IN    New Delhi
CH      Beijing

## UPDATE 

1. Update Column names
- we can assign custom column names using `.columns` property:
```python
    brics.columns = [c.upper() for c in brics.columns]
```
- Rename columns or index labels using `DataFrame.rename()`

```python
    brics.rename(columns={'A': 'CoUnTrY'})
```

2. Update Data
- use `pandas.Series.apply()` to invoke function on values of Series:
```python
    # use `pandas.Series.apply()` to invoke function on values of Series:
    summer_temp = summer_temp.apply(lambda c: c * 1.8 + 32)
```

- use .loc[] indexer, and assign the filtered df to new value:
```python
    brics.loc[brics['population'] > 1000.0, ['capitol']] = 'Crowded'
```

3. Create a new column
- We can do simple arithmetic on existing columns (series), and assign the result to a new column:

```python
    df['score'] = df['midterm'] * 0.4 + df['final'] * 0.6
```


In [ ]:
#UPDATE Column Names

# brics.columns   # Index(['country', 'capitol', 'area', 'population'], dtype='object')

# we can assign custom column names
brics.columns = ['A', 'B', 'C', 'D']

# convert column names to uppercase:
#brics.columns = [c.upper() for c in brics.columns]

# Rename columns or index labels:
brics.rename(columns={'A': 'CoUnTrY'})


CoUnTrY         B       C        D
code                                         
BR          Brazil  BRASILIA   8.516   200.40
RU          Russia    MOSCOW  17.100   143.50
IN           India   CROWDED   3.286  1252.00
CH           China   CROWDED   9.597  1357.00
SA    South Africa  PRETORIA   1.221    52.98

In [ ]:
# UPDATE data
# use .loc[] indexer, and assign the filtered df to new value:
brics.loc[brics['population'] > 1000.0, ['capitol']] = 'Crowded'

brics


country   capitol    area  population
code                                            
BR          Brazil  Brasilia   8.516      200.40
RU          Russia    Moscow  17.100      143.50
IN           India   Crowded   3.286     1252.00
CH           China   Crowded   9.597     1357.00
SA    South Africa  Pretoria   1.221       52.98

In [ ]:
# use `pandas.Series.apply()` to invoke function on values of Series:

# typical summer temperatures (in Celsius) of the G7 capitals
temperatures = {
    'Washington': 27,
    'Ottawa': 23,
    'Tokyo': 27,
    'Paris': 22,
    'Berlin': 22,
    'Rome': 26,
    'London': 20
}

# Create a Pandas Series from the dictionary
summer_temp = pd.Series(temperatures)

# use `pandas.Series.apply()` to invoke function on values of Series:
summer_temp = summer_temp.apply(lambda c: c * 1.8 + 32)
print(summer_temp)

# min function
summer_temp[summer_temp == summer_temp.min()]
# print("Min of series:", summer_temp.min())

# brics['capitol'] = brics['capitol'].apply(lambda x: x.upper())
#print df
# brics

Washington    80.6
Ottawa        73.4
Tokyo         80.6
Paris         71.6
Berlin        71.6
Rome          78.8
London        68.0
dtype: float64


London    68.0
dtype: float64

In [ ]:
# Lets calculate and insert a new column
scores = {
    'student':['Potter', 'Weasley', 'Granger', 'Malfoy'],
    'midterm':[80, 70, 90, 60],
    'final':[85, 75, 95, 65]
}

df = pd.DataFrame(scores)

# do simple arithmetic on existing columns (series), and assign the result to a new column labeled 'score'
df['score'] = df['midterm'] * 0.4 + df['final'] * 0.6
print(df)


   student  midterm  final  score
0   Potter       80     85   83.0
1  Weasley       70     75   73.0
2  Granger       90     95   93.0
3   Malfoy       60     65   63.0


In [36]:
# Create a dictionary with car make, model, and year information
car_data = {
    'Make': ['Toyota', 'Honda', 'Ford', 'Chevrolet', 'Nissan'],
    'Model': ['Camry', 'Civic', 'Mustang', 'Malibu', 'Altima'],
    'Year': [2020, 2021, 2019, 2018, 2022]
}

# Create a pandas dataframe from the dictionary
df = pd.DataFrame(car_data)

# Option 1 - Append using .loc[]
# Note that loc can enlarge its target object
new_car = ['Kia', 'Sorento', 2007]   # Create a list with the new car data
df.loc[len(df)] = new_car            # Append the new row to the dataframe

#Option ?
# new_car = {'Make': 'Volkswagen', 'Model': 'T-ROC', 'Year': 2019}
# df.append()

# Option 2 - create a new dataframe with the additional row and concatenate it with the original dataframe using the concat method.
new_car = {'Make': 'Volkswagen', 'Model': 'T-ROC', 'Year': 2019}  # Create a dictionary with the new car data
new_df = pd.DataFrame(new_car, index=[len(df)])              # Create a new dataframe with the additional row
df = pd.concat([df, new_df], axis=0, ignore_index=True)       # Concatenate the new dataframe with the original dataframe

df

Make    Model  Year
0      Toyota    Camry  2020
1       Honda    Civic  2021
2        Ford  Mustang  2019
3   Chevrolet   Malibu  2018
4      Nissan   Altima  2022
5         Kia  Sorento  2007
6  Volkswagen    T-ROC  2019

## Sorting

1. `DataFrame.sort_index()` sorts object by labels (along an axis).

2. `DataFrame.sort_values()` sorts by the values along either axis.
- Parameters
  - `by`: *str or list of str*: Name or list of names to sort by.
  - `ascending=True`: *bool or list of bool, default True*: Sort ascending vs. descending.

In [68]:
scores = {
    "Student": ["Harry", "Ron", "Hermione", "Draco", "Neville"],
    "Charms": [85, 75, 100, 80, 60],
    "Potions": [80, 85, 95, 82, 65],
    "Muggle Studies": [90, 85, 95, 88, 70],
    "DA Dark Arts": [90, 90, 95, 88, 60]
}

# Note that we are using pandas.DataFrame() constructor.
df = pd.DataFrame(scores)  # default index, 0 to n-1

# Set the 'Student' column as the index
df = df.set_index('Student')    # inplace=False

df.sort_values('DA Dark Arts', ascending=False)

# We can sort on 2 columns, and give different sorting orders to them:
df.sort_values(['DA Dark Arts', 'Charms'], ascending=[True, False]) 

# We can also sort series, a single column
df['Charms'].sort_values()

df.sort_index()  # sort by the index, which is typically the row labels


Charms  Potions  Muggle Studies  DA Dark Arts
Student                                                
Draco         80       82              88            88
Harry         85       80              90            90
Hermione     100       95              95            95
Neville       60       65              70            60
Ron           75       85              85            90

## Analyzing and Exploring

#### Mean, Median, Mode

- When the data contains outliers, the median is often preferred as it is more robust to extreme values. The mean is heavily influenced by outliers and may not be a good representation of the central tendency of the data.
- When the data is categorical or ordinal, the median is not applicable and only the mode can be used.

## Missing Values
- `Series.fillna()`: Fill NA/NaN values using the specified method.

```python
        series = pd.Series([10, np.NaN, -2, None])
        
        print(f"{series.count()= }")    # 2 Note that count returns number of non-NA/null observations in the Series.
        print(f"{series.mean()= }")     # 4.0

        # Replace NaN values with 0
        series = series.fillna(0)
        print(f"{series.mean()= }")     # 2.0
```


### Aggregating
- `Series.describe()`: Generate descriptive statistics.
- `Series.min()`: Return the minimum of the values over the requested axis. If you want the index of the minimum, use `idxmin()`.
- `Series.max()` : Return the maximum of the values over the requested axis. If you want the index of the maximum, use `idxmax()`. 
- `Series.median()`: Return the median of the values over the requested axis.
- `Series.mean()`: Return the mean of the values over the requested axis.
- `Series.mode(dropna=True)`: Return the mode(s) of the Series. The mode is the value that appears most often. There can be multiple modes.
- `Series.sum()`: Return the sum of the values over the requested axis.
- `Series.count()`: Return number of non-NA/null observations in the Series.
- `Series.value_counts()`: Return a Series containing counts of unique values.


In [85]:
# Create a DataFrame for Analyzing and Exploring
scores = {
    "Student": ["Harry", "Ron", "Hermione", "Draco", "Neville"],
    "Charms": [85, 75, 100, 80, 60],
    "Potions": [80, 85, 95, 82, 65],
    "Muggle Studies": [90, 85, 95, 88, 70],
    "DA Dark Arts": [90, 90, 95, 88, 60]
}

# Note that we are using pandas.DataFrame() constructor.
df_scores = pd.DataFrame(scores)  # default index, 0 to n-1

# Set the 'Student' column as the index
df_scores = df_scores.set_index('Student')    # inplace=False

df_scores.describe()  # Generate descriptive statistics.


Charms    Potions  Muggle Studies  DA Dark Arts
count    5.00000   5.000000        5.000000      5.000000
mean    80.00000  81.400000       85.600000     84.600000
std     14.57738  10.830512        9.449868     13.992855
min     60.00000  65.000000       70.000000     60.000000
25%     75.00000  80.000000       85.000000     88.000000
50%     80.00000  82.000000       88.000000     90.000000
75%     85.00000  85.000000       90.000000     90.000000
max    100.00000  95.000000       95.000000     95.000000

In [99]:
# Handle missing values in basic statistics
# Default - ignore. 
# Option - we can replace missing values with another value.
import numpy as np

# In Python, you can represent a null value using the None keyword
series = pd.Series([10, np.NaN, -2, None])
print(f"{series.count()= }")    # 2 Note that count returns number of non-NA/null observations in the Series.
print(f"{series.mean()= }")     # 4.0

# Replace NaN values in 'scores' column with 0
series = series.fillna(0)
print(f"{series.mean()= }")     # 2.0


series.count()= 2
series.mean()= 4.0
series.mean()= 2.0


In [86]:

score_max = df_scores['Charms'].max()   # 100
id_max = df_scores['Charms'].idxmax()   # 'Hermione' (If you want the index of the maximum, use idxmax)

print(f"{df_scores['Charms'].nlargest(3)= }")    # Series
""" 
Hermione    100
Harry        85
Draco        80
Name: Charms, dtype: int64"""

print(f"{score_max= }")
print(f"{id_max= }")    

print(f"{df_scores['Charms'].min()= }")     # 60
print(f"{df_scores['Charms'].idxmin()= }")  # 'Neville' (If you want the index of the maximum, use idxmax)

print(f"{df_scores['Charms'].median()= }")
print(f"{df_scores['Charms'].mean()= }")
print(f"{df_scores['DA Dark Arts'].mode()= }")

print(f"{df_scores['Charms'].sum()= }")     # 400
print(f"{df_scores['Charms'].count()= }")



df_scores['Charms'].nlargest(3)= Student
Hermione    100
Harry        85
Draco        80
Name: Charms, dtype: int64
score_max= 100
id_max= 'Hermione'
df_scores['Charms'].min()= 60
df_scores['Charms'].idxmin()= 'Neville'
df_scores['Charms'].median()= 80.0
df_scores['Charms'].mean()= 80.0
df_scores['DA Dark Arts'].mode()= 0    90
Name: DA Dark Arts, dtype: int64
df_scores['Charms'].sum()= 400
df_scores['Charms'].count()= 5


In [72]:


# Option 1 - Return the first n rows ordered by columns in descending order.
# This method is equivalent to df.sort_values(columns, ascending=False).head(n), but more performant.
df_scores.nlargest(3, 'Charms')

# Option 2 - Return the largest n elements (Series).
df_scores['Charms'].nlargest(3)    # Note that we are calling nlargest() on a series.


Student
Hermione    100
Harry        85
Draco        80
Name: Charms, dtype: int64

In [128]:
df = pd.read_csv("data/sample.csv")

# print(df)

# counts of unique values.
print(df['Is Active'].value_counts())
""" 
Is Active
Yes    5
No     3
Name: count, dtype: int64
"""

# normalize: If True then the object returned will contain the relative frequencies of the unique values.
# counts of unique values.
print(df['Is Active'].value_counts(normalize=True))
""" 
Is Active
Yes    0.625
No     0.375
Name: proportion, dtype: float64"""

print()


Is Active
Yes    5
No     3
Name: count, dtype: int64
Is Active
Yes    0.625
No     0.375
Name: proportion, dtype: float64



## GROUP BY


In [163]:
# GROUP BY
df = random_users   # random data created at the top of the page.
# print(df)

df['Is Active'].value_counts()
""" 
Is Active
No     502
Yes    498
Name: count, dtype: int64
"""

df.groupby(['Country'])['Is Active'].value_counts()
""" 
Country    Is Active
Australia  No           60
           Yes          44
Canada     No           46
           Yes          44
                       ...
USA        Yes          46
           No           37
Name: count, dtype: int64
"""

df['Salary'].median()

median_salary_by_country = df.groupby(['Country'])['Salary'].median()
""" 
Country
Australia    578.0
Canada       484.5
China        543.0
France       496.0
Germany      491.0
India        514.0
Japan        424.0
Russia       539.0
UK           569.0
USA          557.0
Name: Salary, dtype: float64
"""
print(median_salary_by_country)


Country
Australia    578.0
Canada       484.5
China        543.0
France       496.0
Germany      491.0
India        514.0
Japan        424.0
Russia       539.0
UK           569.0
USA          557.0
Name: Salary, dtype: float64
